### DISCUSSION TITLE
Virtual List for Foldable/Iterable

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#8558] [Ideas]
pattern match on any Foldable/Iterable as though it is a list, without conversion the whole thing to a list

```
enum VirtualList[t, l] {
    case Cons(t, l)
    case Nil
}

trait Foldable[t: Type -> Type] {
    pub def vList(l: t[a]): VirtualList[a, t[a]) \ .. = ..
}
```

**Comment by JonathanStarup** [comment]
maybe `Option[(x, t[a])]` is less alien?

**Comment by magnus-madsen** [comment]
https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/view_patterns.html
